File: src/main/java/com/keycloak/reactive/pets/PetService.java
package com.keycloak.reactive.pets;

import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@RequiredArgsConstructor
@Service
public class PetService{
    @Autowired
    private final PetRepository petRepository;

    public Mono<PetDTO> getPetById(String id) {
        return petRepository
            .findById(id)
            .switchIfEmpty(Mono.error(new Exception("Pet not found")))
            .flatMap(pet -> Mono.just(PetDTO.fromPet(pet)));
    }

}

File: src/main/kotlin/com/keycloak/reactive/pets/PetService.kt
package com.keycloak.reactive.pets

import org.springframework.stereotype.Service
import reactor.core.publisher.Mono

@Service
class PetService(
    private val petRepository: PetRepository
) {

    fun getPetById(id: String): Mono<PetDTO> {
        return petRepository
            .findById(id)
            .flatMap { Mono.just(PetDTO(it)) }
    }

}